### ISSUE TITLE
Removing need to hash password before sending to wesnothd/campaignd from the client

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#5132]
A problem that's come up with the current implementation of logging into the official MP server with the forum username and password is that the wesnoth client needs to know how to hash the password it's sending, since all communication between wesnothd/campaignd is unencrypted.  This is an issue, since the hashing algorithm used by the forum software tends to change between releases which then requires altering which hashing algorithm the forums software uses so that the hashed password sent by the wesnoth client can be compared to the hashed password stored in the forum database.

For example, if the wesnoth client only knows how to hash passwords with MD5, and a forum upgrade changes the default password hashing to use bcrypt, then obviously the MD5 hash and the bcrypt hash aren't going to match.  This has come up again recently with phpbb now changing from preferring bcrypt to preferring argon2.

Short version: the wesnoth client shouldn't need to know how to create a hash that matches the phpbb forum software.

---

Based on some previous discussions, it sounds like the best solution would be to OpenSSL(?) to encrypt communication done between wesnothd/campaignd and the client, which would remove the need to have the password hashed by the client.

**Comment by loonycyborg** [comment]
Asio has built-in support for ssl: https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/overview/ssl.html
Most likely it will be matter of just using that extra api. But also matter of certificates requires research. From what I gather there is no general way to properly make use of system certificate authorities with asio in cross-platform manner atm. Perhaps we can supply our own certs since protocol that will be carried by the ssl layer is specific to wesnoth too?

**Comment by Pentarctagon** [comment]
If we provide our own certs, one issue would be that they would need to stay valid for the life of the current stable branch and the next stable branch.

**Comment by Pentarctagon** [comment]
@loonycyborg given the issues this causes when upgrading the forums, is it something you'd be able to investigate sometime during the 1.15 development cycle?

**Comment by loonycyborg** [comment]
Maybe. Still would like to hear more comments about this from other people. From @GregoryLundberg for example.

**Comment by Pentarctagon** [comment]
I don't think he's still around.  Regardless though, I'll probably give it a try in the next couple months and see if maybe I can get it to work.
